### ISSUE TITLE
"Observe DOM changes" in Visibility Trigger causes JS error on load

### ISSUE THREAD

**Issue by micvm** [matomo-org/tag-manager#481]
I am using matomo 4.9.1 and I wonder if I am doing something wrong or if there is a bug in the tag manager. If I check the option "Observe DOM Changes" in a  Visibility trigger, regardless of which element I want to observe (i.e. whether the element exists or not), I get the following error in the container-js file upon loading the page:

```
Uncaught TypeError: addedNodes is undefined
    mutationObserverCallback https://myurl/matomo/js/container_a9N6jAO3.js:227
    observerMutation https://myurl/matomo/js/container_a9N6jAO3.js:225
    setTimeout handler*Templates.ElementVisibilityTrigger/setMutationObserver/Templates.ElementVisibilityTrigger/observerMutation< https://myurl/matomo/js/container_a9N6jAO3.js:225
    ElementVisibilityTrigger https://myurl/matomo/js/container_a9N6jAO3.js:224
    onLoad https://myurl/matomo/js/container_a9N6jAO3.js:24
    setUp https://myurl/matomo/js/container_a9N6jAO3.js:237
    setUp https://myurl/matomo/js/container_a9N6jAO3.js:28
    run https://myurl/matomo/js/container_a9N6jAO3.js:35
    addContainer https://myurl/matomo/js/container_a9N6jAO3.js:36
    <anonymous> https://myurl/matomo/js/container_a9N6jAO3.js:244
    <anonymous> https://myurl/matomo/js/container_a9N6jAO3.js:244
    <anonymous> https://myurl/matomo/js/container_a9N6jAO3.js:245
container_a9N6jAO3.js:formatted:6711
```
The corresponding line in the code of the container-js (cf. line `addedNodes.forEach`):
```js
function mutationObserverCallback(mutationsList, triggerEvent) {
          var domElements = TagManager.dom.bySelector(selectors);
          for (var index in mutationsList) {
            var mutation = mutationsList[index];
            var addedNodes = mutation.addedNodes;
            if (mutation.type === 'attributes') {
              addedNodes = [
                mutation.target
              ];
            }
            addedNodes.forEach(function (node) {
              domElements.forEach(function (element) {
                if (node.contains(element)) {
                  if (blockTrigger || (onlyOncePerElement && isNodeEventTriggered(element))) {
                    return;
                  }
```
If that matters: I tried to select by ElementID as well as with CSS selector and that behavior does not change.

**Comment by AltamashShaikh** [comment]
@micvm what's your browser version?

**Comment by micvm** [comment]
Happens with Firefox 99 (64-bit) as well as with Chromium (Version 101.0.4951.64 (Official Build) snap (64-bit)), both on Ubuntu. I am not a JS expert but when the JS debugger stops because of the exception the index variable has the value `"each"` which looks somehow weird because if I set a breakpoint within the loop before the exception `index` has as expected the values 0,1,2,... as it should.

**Comment by AltamashShaikh** [comment]
@micvm possible to share any URL where I can check this issue ?

**Comment by micvm** [comment]
I've sent you the URL of the corresponding page via email.

**Comment by AltamashShaikh** [comment]
@micvm 
I see `76e6a1758.combined.js.h-1905639948.pack` has some each function which is adding an extra element with `each` as index in the mutation list which is a function and hence we cannot parse the node.
We can add a check for array at most

**Comment by AltamashShaikh** [comment]
@micvm can you replace Template/Trigger/ElementVisibilityTrigger.web.js with this [file](https://github.com/matomo-org/tag-manager/blob/0375f3b641ed937cbe360f610e407c7df45758d4/Template/Trigger/ElementVisibilityTrigger.web.js)

After replacing the file, you need to run `./console tagmanager:regenerate-released-containers` and check if the error persists or not

**Comment by micvm** [comment]
Hi there! Awesome, the changed version works perfectly. I wonder where precisely this "each" is added... If you have a clue, very appreciated as it looks like something should not happen :) Thanks for the fix!!

**Comment by AltamashShaikh** [comment]
@micvm The fix was merged, but it will be released with Matomo 4.11 which will take atleast a month to release, since Matomo 4.10 was released today if you update it the current fix would go away and you might need to reapply the patch after upgrade.

**Comment by micvm** [comment]
No problem. Thanks for letting me know and the quick fix!
